### PR TITLE
Detect public functions to be declared as external

### PR DESCRIPTION
**Pull request by samparsky**
Fixes issue #29 

@montyly Can you help review

### CODE REVIEW

**Comment by montyly** [comment]
Thanks for your contribution!

- `INTERNAL_DYNAMIC_CALL` `.function` returns a `Variable`, and not a `Function`; this is because it's a dynamic call which may point to many different functions (and the set of potential destinations is not yet computed). Finding all the destinations is out-of-scope of this detector, I think we can simply discard the detector if we find any `INTERNAL_DYNAMIC_CALL` in the contract

- If the function is called through a `HighLevelCall`, it should not be considerd as non-potential external function, a `HighLevelCall` is an external call to another contract.

- Ather some thinking, I think `def detect_public(self, contract):` could just return `contract.all_functions_called`, which returns all the functions reachable by the contract

File: slither/detectors/functions/external_function.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        """ Returns a list of InternallCall, InternalDynamicCall, SolidityCall, HighLevelCall
+            calls made in a function
+
+        Returns:
+            (list): List of all InternallCall, InternalDynamicCall, SolidityCall, HighLevelCall
+        """
+        result = []
+        for node in func.nodes:
+            for ir in node.irs:
+                if isinstance(ir, ( InternalCall, InternalDynamicCall, HighLevelCall, SolidityCall )):
```
Comment (disconnect3d) [inline slither/detectors/functions/external_function.py]:
```suggestion
                if isinstance(ir, (InternalCall, InternalDynamicCall, HighLevelCall, SolidityCall)):
```

**Comment by samparsky** [comment]
@monthly Can you clarify further what you mean by this?
>  I think we can simply discard the detector if we find any INTERNAL_DYNAMIC_CALL in the contract

**Comment by montyly** [comment]
Without an additional analysis, you can't know the set of potential functions executed through an `InternalDynamicCall` as its computed at runtime (through its possible to compute it statically, slither doesn't do it yet).

Consider the following example:
```
contract C{
    
    function () returns(uint) ptr;
    
    
    function set_test1() external{
        ptr = test1;
    }
    
    function set_test2() external{
        ptr = test2;
    }
    
    function test1() public returns(uint){
        return 1;
    }
    
    function test2() public returns(uint){
        return 2;
    }
    
    function test3() public returns(uint){
        return 3;
    }
    
    function exec() external returns(uint){
        return ptr();
    }
    
}
```

Here you have three functions that are `public`, and are candidates to become `external`. But only `test3()` can, because `ptr` is never assigned to `test3`

With the IR we can compute the set of function that can be associated to `ptr`, though I think it's simpler to just not analyze a contract if it has a `InternalDynamicCall` for this detector

**Comment by samparsky** [comment]
@montyly I am having issues compiling the example contract you gave above. The slithIR  gives the error below

```
 raise VariableNotFound('Variable not found: {}'.format(var_name))
slither.solc_parsing.expressions.expression_parsing.VariableNotFound: Variable not found: ptr()
```

**Comment by samparsky** [comment]
@montyly Its working okay after doing a rebase on the master branch

**Comment by samparsky** [comment]
@montyly Kindly review,
I have added that a contract should be ignored if it contains an InternalDynamicCall

**Pull request merged**